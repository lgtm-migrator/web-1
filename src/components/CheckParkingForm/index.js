import React from "react"
import Button from "@material-ui/core/Button"
import TextField from "@material-ui/core/TextField"

import Fieldset from "../Fieldset"
import Form from "../Form"

export default ({ loading, onCarPlateChange, onSubmit }) => (
  <Form onSubmit={onSubmit}>
    <Fieldset variant="borderNone">
      <TextField
        id="car_plate"
        label="Placa do carro"
        InputProps={{
          "aria-required": true,
        }}
        inputProps={{
          maxLength: 7,
          minLength: 7,
        }}
        name="car_plate"
        onChange={onCarPlateChange}
        placeholder="Letras e nÃºmeros"
        required={true}
        size="small"
        type="search"
        variant="outlined"
      />
      <Button
        disabled={loading}
        size="medium"
        type="submit"
        variant="contained"
      >
        Verificar
      </Button>
    </Fieldset>
  </Form>
)
