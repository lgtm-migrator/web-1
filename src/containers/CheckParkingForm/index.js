import React, { useEffect, useReducer, useState } from "react"
import CheckParkingForm from "../../components/CheckParkingForm"
import { getParkingsByCarPlate } from "../../services/parkings"
import { useSetReportedParkingsState } from "../../hooks/useReportedParkingsState"
import useLoadingState from "../../hooks/useLoadingState"
import { getHealthCheckResponse } from "../../services/healthcheck"
import Alert from "@material-ui/lab/Alert"
import AlertTitle from "@material-ui/lab/AlertTitle"
import Button from "@material-ui/core/Button"
import Link from "@material-ui/core/Link"

export default ({ carPlate }) => {
  const [loading, setLoading] = useLoadingState()
  const [apiError, setApiError] = useState(false)

  const initialState = {
    carPlate: "",
    loading: false,
    success: null,
  }

  const reducer = (state, action) => {
    switch (action.type) {
      case "api_request":
        return {
          loading: true,
        }
      case "api_success":
        return {
          loading: false,
          success: true,
        }
      case "api_error":
        return {
          loading: false,
          success: false,
        }
      case "not_found_error":
        return {
          loading: false,
          success: true,
        }
      default:
        break
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState)
  const [stateCarPlate, setCarPlate] = useState(carPlate)
  const setReportedParkingsState = useSetReportedParkingsState()

  useEffect(() => {
    window.history.pushState(
      { carPlate },
      null,
      `/consultar?car_plate=${carPlate}`
    )
  })

  const handleCarPlateChange = (event) => {
    setCarPlate(event.target.value)
  }

  const handleApiResponse = (response) => {
    dispatch({ type: "api_success" })

    const {
      data: { parkings },
    } = response.data
    setReportedParkingsState(parkings)
  }

  const handleApiError = (error) => {
    const isNotFoundError =
      typeof error.response !== "undefined" && error.response.status === 404
    if (isNotFoundError) {
      dispatch({ type: "not_found_error" })
      setReportedParkingsState([])
      return
    }
    dispatch({ type: "api_error" })
    setReportedParkingsState(null)
  }

  const handleSubmit = async (event) => {
    event.preventDefault()
    setLoading(true)

    dispatch({ type: "api_request" })

    try {
      const response = await getParkingsByCarPlate(stateCarPlate)
      handleApiResponse(response)
    } catch (error) {
      handleApiError(error)
    }
  }

  const handleHealthCheck = () => {
    getHealthCheckResponse()
      .then(() => {
        setLoading(false)
      })
      .catch(() => {
        setApiError(true)
      })
  }

  useEffect(() => {
    handleHealthCheck()
  }, [])

  return (
    <>
      {apiError ? (
        <Alert severity={"error"}>
          <AlertTitle>
            Ops! Não é possível consultar denúncias agora.
          </AlertTitle>
          Nosso sistema de consulta de denúncias parece estar fora do ar.
          <br />
          Você pode{" "}
          <Button color="inherit" component={Link} onClick={handleHealthCheck}>
            tentar novamente
          </Button>{" "}
          em alguns instantes.
        </Alert>
      ) : (
        <CheckParkingForm
          // carPlate={stateCarPlate}
          loading={loading}
          // onCarPlateChange={handleCarPlateChange}
          onSubmit={handleSubmit}
        />
      )}
    </>
  )
}
