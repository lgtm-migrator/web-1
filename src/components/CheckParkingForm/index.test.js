import React from "react"
import renderer from "react-test-renderer"
import CheckParkingForm from "."

describe("<CheckParkingForm /> component", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<CheckParkingForm />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
