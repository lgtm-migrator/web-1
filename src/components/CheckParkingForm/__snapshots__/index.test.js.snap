// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckParkingForm /> component should render correctly 1`] = `
<form
  className="MuiBox-root MuiBox-root-2"
  data-name="form"
  id="form"
  name="form"
>
  <fieldset
    className="MuiBox-root MuiBox-root-6 makeStyles-root-5 makeStyles-borderNone-3"
  >
    <div
      autoCorrect="off"
      className="MuiFormControl-root MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
        data-shrink={false}
        htmlFor="car_plate"
        id="car_plate-label"
      >
        Placa do carro
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          aria-required={true}
          autoCapitalize="characters"
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input makeStyles-input-7 MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch"
          data-name="car_plate"
          disabled={false}
          id="car_plate"
          inputMode="verbatim"
          maxLength={7}
          minLength={7}
          name="car_plate"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          pattern="[a-zA-Z]{3}[0-9]([a-zA-Z]|[0-9])[0-9]{2}"
          placeholder="Letras e números"
          required={true}
          type="search"
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Verificar
      </span>
    </button>
  </fieldset>
</form>
`;
